### ISSUE TITLE
Look into performance improvements for crop

### ISSUE THREAD

**Issue by djohnson729** [ngageoint/mrgeo#281]

**Comment by djohnson729** [comment]
The modification for improving crop performance involves configuring the input format to perform filtering to prevent reading all of the tiles outside of the crop region. This can only be done if the input raster to the crop is a stored image (as opposed to the result of a map algebra expression).

I am changing the crop map op so that it checks to see if its raster input is a MrsPyramidMapOp. If it is,it sets the crop bounds in that map op (after cloning it so as not to affect any other map op's that use the same raster input). If the MrsPyramidMapOp has bounds assigned to it (as described above), it in turn configures the Hadoop input format with those bounds. The input format has logic to limit the returned splits (and records) to only include tiles within those bounds, thus preventing the reading of all tiles in the image unnecessarily.

Note that if the input raster passed to crop is not a reference to a stored image (e.g. it is another map algebra expression), it will use the existing Spark filtering because that's all it can do.

**Comment by djohnson729** [comment]
In my testing within AWS EMR, a crop operation on a worldwide image ran in 22 minutes, 12 seconds with the old code, and now runs in 1 minute, 38 seconds. And the tobler friction surface computation (which required two crop operations ) that ran in approximately 50 minutes, now runs in 2 minutes and 42 seconds.

**Comment by drew-bower** [comment]
A couple questions-

What was the crop operation (how big)?

What change was made to speed this up?

**Comment by djohnson729** [comment]
I cropped a worldwide image down to 2 degrees square

In a nutshell, I made changes to perform the filtering at the data input level rather than waiting until the data is flowing through Spark. That prevents ever having to push all the unneeded tiles through the system. There is more detail on this in the second paragraph of my earlier comment along with conditions under which is still has to perform the Spark filtering.
